src/sat.cc

```diff
@@ -126,122 +127,98 @@ bool Sat::CheckEnvironment() {
+                           "--cpu_freq_threshold be set to a positive value."});
       return false;
     } else if (cpu_freq_round_ < 0) {
-      logprintf(0,
-                "Process Error: The --cpu_freq_round option must be greater"
-                " than or equal to zero. A value of zero means no rounding.\n");
-      bad_status();
+      setup_step.AddError(Error{
+          .symptom = kProcessError,
+          .message = "The --cpu_freq_round option must be greater than or "
+                     "equal to zero. A value of zero means no rounding."});
       return false;
     }
   }
 
   // Use all CPUs if nothing is specified.
   if (memory_threads_ == -1) {
     memory_threads_ = os_->num_cpus();
-    logprintf(7, "Log: Defaulting to %d copy threads\n", memory_threads_);
+    setup_step.AddLog(Log{
+        .severity = LogSeverity::kDebug,
+        .message = absl::StrFormat(
+            "Defaulting to using %d memory copy threads (same number as there "
+            "are CPU cores)",
+            memory_threads_)});
   }
 
   // Use all memory if no size is specified.
-  if (size_mb_ == 0) size_mb_ = os_->FindFreeMemSize() / kMegabyte;
+  if (size_mb_ == 0) size_mb_ = os_->FindFreeMemSize(setup_step) / kMegabyte;
   size_ = static_cast<int64>(size_mb_) * kMegabyte;
 
-  // Autodetect file locations.
-  if (findfiles_ && (file_threads_ == 0)) {
-    // Get a space separated sting of disk locations.
-    list<string> locations = os_->FindFileDevices();
-
-    // Extract each one.
-    while (!locations.empty()) {
-      // Copy and remove the disk name.
-      string disk = locations.back();
-      locations.pop_back();
-
-      logprintf(12, "Log: disk at %s\n", disk.c_str());
-      file_threads_++;
-      filename_.push_back(disk + "/sat_disk.a");
-      file_threads_++;
-      filename_.push_back(disk + "/sat_disk.b");
-    }
-  }
-
   // We'd better have some memory by this point.
   if (size_ < 1) {
-    logprintf(0, "Process Error: No memory found to test.\n");
-    bad_status();
+    setup_step.AddError(
+        Error{.symptom = kProcessError,
+              .message = "No memory found to test on the system."});
     return false;
   }
 
   if (tag_mode_ &&
       ((file_threads_ > 0) || (disk_threads_ > 0) || (net_threads_ > 0))) {
-    logprintf(0,
-              "Process Error: Memory tag mode incompatible "
-              "with disk/network DMA.\n");
-    bad_status();
+    setup_step.AddError(Error{
+        .symptom = kProcessError,
+        .message =
+            "Memory tag mode is incompatible with disk and network testing."});
     return false;
   }
 
   // If platform is 32 bit Xeon, floor memory size to multiple of 4.
   if (address_mode_ == 32) {
     size_mb_ = (size_mb_ / 4) * 4;
     size_ = size_mb_ * kMegabyte;
-    logprintf(1, "Log: Flooring memory allocation to multiple of 4: %lldMB\n",
-              size_mb_);
-  }
-
-  // Check if this system is on the whitelist for supported systems.
```

Comment:
Same here. I don't know why unsupported os won't be handled but I will trust you.

Reply:
Again, not implemented in this code. It just calls an empty function that always returns true.